Use Helm to install Calico CNI in e2e tests instead of ClusterResourceSets

CecileRobertMichon opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**

/kind cleanup

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:  

This PR replaces the existing ClusterResourceSet-dependent solution for automating calico CNI delivery onto test clusters (and reference cluster templates) with a helm solution. For the private cluster scenario, we're keeping the CRS-dependent solution as a practical way to install the CNI from the management cluster (we have no direct access to the apiserver to helm install gestures and we don't yet have a E2E solution to indirectly access the apiserver in such scenarios).

This work is a continuation of @jackfrancis' work in #2334.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2063
Contributes to https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2179

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Use Helm to install Calico CNI
```

jackfrancis commented:
oooh 👀👀👀👀👀👀👀

jackfrancis commented on test/e2e/helpers.go:
```diff
@@ -879,6 +879,7 @@ func InstallHelmChart(ctx context.Context, input clusterctl.ApplyClusterTemplate
 		p := helmGetter.All(settings)
 		valueOpts := &helmVals.Options{}
 		valueOpts.Values = values
+		valueOpts.ValueFiles = valueFiles
```
I assume the helm api handles an empty array here gracefully? And/or does the same if we include data in both `Values` and `ValueFiles` properties?

(Wondering if we need to do some input validation sanity-checking...)

CecileRobertMichon replied:
Yes `MergeValues` will iterate over each `[]string` of values and merge them https://github.com/helm/helm/blob/main/pkg/cli/values/options.go#L41

Jont828 commented on test/e2e/helpers.go:
```diff
@@ -858,29 +859,39 @@ func getPodLogs(ctx context.Context, clientset *kubernetes.Clientset, pod corev1
 	}
 	i.ReleaseName = releaseName
-	i.Namespace = ns
+	i.Namespace = namespace
```
The Helm install client has a flag to create a namespace if it doesn't exist. You can add this below.
```
i.CreateNamespace = true
```

CecileRobertMichon replied:
thank you, I didn't know, that's great!!

Jont828 commented on test/e2e/cni.go:
```diff
@@ -0,0 +1,109 @@
+	if ipv6 != "" && ipv4 != "" {
+		// dual-stack
+		values = &helmVals.Options{
+			StringValues: []string{fmt.Sprintf("installation.calicoNetwork.ipPools[0].cidr=%s", ipv4), fmt.Sprintf("installation.calicoNetwork.ipPools.ipPools[1].cidr=%s", ipv6)},
```
Do we plan on using the ipPool CIDRs specified in the different values.yaml files? Or is it just there as a default?

CecileRobertMichon replied:
The ipPool CIDRs in the values.yaml are there as default (they match the default values in cluster-template.yaml) so the file can be used outside of e2e tests, for example in the quickstart instructions

Jont828 commented on test/e2e/cni.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func getCalicoValues(cidrBlocks []string) *helmVals.Options {
+	var ipv6, ipv4 string
```
It looks like this would overshadow the const `ipv4` and `ipv6` declared at the top of the file and are used to specify CIDR blocks.
```suggestion
	var ipv6CidrBlock, ipv4CidrBlock string
```

Jont828 commented on templates/addons/calico-dual-stack/values.yaml:
```diff
@@ -0,0 +1,19 @@
+    mtu: 1350
+    ipPools:
+    - blockSize: 26
+      cidr: 10.244.0.0/16
```
Would there be a situation where a pod CIDR list has multiple IPv4 address blocks? If so, how would we set this field? Would it be the union of the two blocks?

CecileRobertMichon replied:
I don't think that's possible but let me do more research on that to verify

CecileRobertMichon replied:
confirmed. CAPI is also adding validation to enforce this: https://github.com/kubernetes-sigs/cluster-api/pull/7420

CecileRobertMichon commented:
/retest

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e

CecileRobertMichon commented:
/test ls

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-apiversion-upgrade

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance
/test pull-cluster-api-provider-azure-apiversion-upgrade

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance
/test pull-cluster-api-provider-azure-apiversion-upgrade

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance
/test pull-cluster-api-provider-azure-apiversion-upgrade

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented on go.mod:
```diff
@@ -44,7 +44,7 @@ require (
 	k8s.io/utils v0.0.0-20220210201930-3a6ce19ff2f9
 	sigs.k8s.io/cluster-api v1.2.0
-	sigs.k8s.io/cluster-api/test v1.2.0
+	sigs.k8s.io/cluster-api/test v1.2.1-0.20220725220710-f3c6c6de4c40
```
TODO: update this to 1.2.1 once it is released

CecileRobertMichon replied:
done

CecileRobertMichon commented on go.mod:
```diff
@@ -44,7 +44,7 @@ require (
 	k8s.io/utils v0.0.0-20220210201930-3a6ce19ff2f9
 	sigs.k8s.io/cluster-api v1.2.0
-	sigs.k8s.io/cluster-api/test v1.2.0
+	sigs.k8s.io/cluster-api/test v1.2.1-0.20220725220710-f3c6c6de4c40
```
TODO: update this to 1.2.1 once it is released

CecileRobertMichon replied:
done

CecileRobertMichon commented on docs/book/src/topics/addons.md:
```diff
@@ -0,0 +1,158 @@
+
+For more information, see the [official Calico documentation](https://projectcalico.docs.tigera.io/getting-started/kubernetes/helm).
+
+## Flannel
```
note for reviewers: this section was copied as-is from flannel.md

CecileRobertMichon commented on docs/book/src/topics/addons.md:
````diff
@@ -0,0 +1,158 @@
+kubectl apply -f kube-flannel.yml
+```
+
+# External Cloud Provider
````
note for reviewers: this section was copied as-is from cloud-provider-config.md

CecileRobertMichon commented on docs/book/src/topics/addons.md:
```diff
@@ -0,0 +1,158 @@
+
+<h1> Note </h1>
+
+For Windows nodes, you also need to copy the kubeadm-config configmap to the calico-system namespace so the calico-node-windows Daemonset can find it:
```
@marosset I'm concerned that this will break upstream tests out there that rely on the CAPZ windows calico resources, or that it will confuse users. It works as a workaround for our tests but it's not a great long-term solution. It would be much better if we can figure out a way to solve https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/hostprocess/calico/install/calico-install.ps1#L13 and not rely on the kubeadm-config configmap.

jackfrancis replied:
I assume we're good with this @marosset, correct?

CecileRobertMichon commented:
This is overall ready for review, added some docs and cleaned up commits.

The only thing blocking at this point is the self-hosted test which is failing, I still need to figure out why. In the meantime we can start reviewing and I'll keep this on hold so it doesn't merge.

/hold

/assign @jackfrancis @Jont828 @sonasingh46

CecileRobertMichon commented:
The self-hosted test is failing due to https://github.com/projectcalico/calico/issues/3145 which we worked around in the manifest by adding `featureDetectOverride: ChecksumOffloadBroken=true` to Felix configuration in manifest. It's not as easy to workaround with Helm chart as it doesn't allow direct configuration of the FelixConfiguration CRD. 

Trying to understand why the bug still happens in the first place since the root cause kernel bug has been fixed (https://github.com/torvalds/linux/commit/ea64d8d6c675c0bb712689b13810301de9d8f77a) since April 2020.

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-capi-e2e

CecileRobertMichon commented:
/retest

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
/retest

CecileRobertMichon commented:
Current status of this:

1) needs rebase
2) needs reviews
3) two things making me hesitate to move this forward - requires workaround for Windows kube-proxy and workaround for FelixConfiguration because of the ChecksumOffloadBroken kernel bug. Not blockers per se since they both have workarounds but this is a regression in UX.

jackfrancis commented:
Are there any open calico issues tracking those workarounds?

Just a thought: now that we've established working proficiency with doing this via helm generally, would it make sense to timebox a cilium spike? Perhaps this has a chart that requires less hacking on our part?:

https://github.com/cilium/charts

AFAIK we don't have any particular reason (other than historical momentum) to prefer one standard IPAM CNI solution over another, my understanding is that cilium is functionally equivalent to calico in terms of IPAM.

CecileRobertMichon commented:
> Are there any open calico issues tracking those workarounds?

Yes for the calico FelixConfiguration: https://github.com/projectcalico/calico/issues/4727

For the Windows kube-proxy relying on the kubeadm configmap there is a [TODO](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/hostprocess/calico/install/calico-install.ps1#L13) but I'm not aware of an issue tracking it - @jsturtevant / @marosset do you know?

> Just a thought: now that we've established working proficiency with doing this via helm generally, would it make sense to timebox a cilium spike? Perhaps this has a chart that requires less hacking on our part?

Yes that's a great idea! I was actually going to investigate because of #2723 (which makes it seem like it doesn't work out of the box either). I think they aren't mutually exclusive though, I would advocate that we move this PR forward without removing the existing Calico install yaml used in the quickstart, but at least start by replacing ClusterResourceSets used in the e2e tests. And look into adding side by side e2e tests with Cilium in parallel. If we find that one is better than the other UX wise and they are both equivalent performance wise we can revisit what we use as "default" in docs and quickstart.

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon reviewed (commented):
/assign @jackfrancis @marosset 

This is officially ready for review

/hold cancel

jsturtevant commented:
We don't have an issue (We probably should as this has come up recently now with HostProcess containers going to Stable in 1.26). There are details on the changes required in https://docs.google.com/document/d/1A6Gkyx5EvL86Z4L2P4sxnk9HQjdRgnLXvBg6tIMDB3s/edit# which was presented to sig-windows

CecileRobertMichon commented on docs/book/src/topics/addons.md:
```diff
@@ -0,0 +1,162 @@
+
+Some of the instructions below use [Helm](https://helm.sh) to install the addons. If you're not familiar with using Helm to manage Kubernetes applications as packages, there's lots of good [Helm documentation on the official website](https://helm.sh/docs/). You can install Helm by following the [official instructions](https://helm.sh/docs/intro/install/).
+
+## Calico
```
Going to work on adding a Cilium section here as a follow up, once I look into #2723

jackfrancis replied:
ACK

CecileRobertMichon commented on templates/addons/windows/calico/calico.yaml:
```diff
@@ -11,7 +11,7 @@ kind: ConfigMap
 apiVersion: v1
 metadata:
   name: calico-static-rules
-  namespace: kube-system
```
@marosset this is important to note: because the Calico Helm chart puts everything in `calico-system` namespace, it requires the windows components to be in the same namespace which is why I changed all these manifests. We might want to check if these are being used anywhere outside CAPZ so this doesn't break users / tests when it merges...

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
/test ls

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-csi-migration

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-csi-migration
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
This is still ready for review

/lifecycle active

Preference for not merging today though, given that we're releasing tomorrow and this is a huge change.

Once Calico 1.25 is out we'll be able to revert the last commit for the FelixConfiguration workaround.

CecileRobertMichon commented:
Need to look into why tests are failing...

CecileRobertMichon commented:
/retest

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-apiversion-upgrade

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-apiversion-upgrade

CecileRobertMichon commented:
@jackfrancis would you be able to give this one another review? Would be great to get it merge soon-ish so I can stop rebasing and we can have sufficient mileage before the v1.7 release 

(tests are failing due to ongoing docker [outage](https://status.docker.com/))

jackfrancis commented on Tiltfile:
```diff
@@ -350,8 +346,20 @@ def deploy_worker_templates(template, substitutions):
+        calico_values = "./templates/addons/calico-dual-stack/values.yaml"
+    else:
+        calico_values = "./templates/addons/calico/values.yaml"
+    flavor_cmd += "; " + helm_cmd + " repo add projectcalico https://projectcalico.docs.tigera.io/charts; " + helm_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig install calico projectcalico/tigera-operator -f " + calico_values + " --namespace tigera-operator --create-namespace; kubectl --kubeconfig ./${CLUSTER_NAME}.kubeconfig apply -f ./templates/addons/calico/felix-override.yaml"
```
should we wait for some outcome from the helm install (like the existence of a CRD, for example) before applying the override? I assume there is some serialization required there?

CecileRobertMichon replied:
that's a good point, from my observations as soon as the Helm chart is applied the CRD exists but let me make sure there isn't a possible race

jackfrancis replied:
Thinking about this more... I think the helm CLI exits success when the apiserver has admitted the full set of resources. Because the CRDs we're talking about don't have any interesting state transitions, we should be able to reliably update `default` of type `FelixConfiguration` w/ `spec.featureDetectOverride: "ChecksumOffloadBroken=true"`.

But we should be aware that we don't fully understand all the side-effects that the set of k8s components that calico is installing via helm. It's _possible_ there's some conflicting update to that same CRD resource that happens during "calico bootstrap", so if we see weird results in our E2E tests we can check here.

CecileRobertMichon replied:
Agreed. Also not that this workaround can be reverted as soon as Calico v1.25 comes out, the fix has already been merged, just waiting for the release

jackfrancis commented on docs/book/src/topics/addons.md:
````diff
@@ -0,0 +1,162 @@
+kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-azure/main/templates/addons/calico/felix-override.yaml
+```
+
+### For IPv4/IPv6 dual-stack clusters
````
I think I'd prefer referring to this as something like "dual-stack (IPv4 + IPv6)"

jackfrancis commented on docs/book/src/topics/addons.md:
````diff
@@ -0,0 +1,162 @@
+
+For Windows nodes, you also need to copy the kubeadm-config configmap to the calico-system namespace so the calico-node-windows Daemonset can find it:
+
+```bash
````
yikes, is there no automated way to do this?

jackfrancis replied:
The only thing I can think of is doing this automatically in the capz control loop. That's not better. :/

CecileRobertMichon replied:
@marosset @jsturtevant are there still plans to fix this for 1.26?

marosset replied:
We need to wait until containerd v1.7 drops :(

marosset replied:
Let's address this as a separate change

jackfrancis replied:
sgtm

jsturtevant replied:
This is due to using the calico operator.  The images were build with the release manifests they publish on github.  We could probably clean the images to remove this but there are better ways to handle a lot of this once containerd 1.7 lands.

CecileRobertMichon commented:
/milestone v1.7

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-capi-e2e
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/test pull-cluster-api-provider-azure-e2e-workload-upgrade
/test pull-cluster-api-provider-azure-conformance

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-csi-migration

CecileRobertMichon commented:
Looks like some flakes on machine provisioning

/retest

jackfrancis commented on scripts/ci-entrypoint.sh:
```diff
@@ -131,6 +131,59 @@ create_cluster() {
+
+    # Add FeatureOverride for ChecksumOffloadBroken in FelixConfiguration.
+    # This is the recommended workaround for https://github.com/projectcalico/calico/issues/3145.
+    "${KUBECTL}" apply -f "${REPO_ROOT}"/templates/addons/calico/felix-override.yaml
```
Do we need to do the Windows foo here as well for Windows scenarios that use ci-entrypoint?

jackfrancis commented:
/lgtm
/approve

take some well deserved 🏖️ @CecileRobertMichon 😎
